internal/dkg/actions_passive.go

```diff
@@ -2,130 +2,133 @@ package dkg
 
-	err := d.executeAction("abort DKG", abort.Metadata.BeaconID, func(_ *drand.Participant, current *DBState) (*DBState, error) {
-		err := d.verifyMessage("StartAbort", abort.Metadata, termsFromState(current))
-		if err != nil {
-			return nil, err
-		}
-		return current.Aborted()
-	})
+	// if we're in the DKG protocol phase, we automatically broadcast it
+	if packet.GetDkg() != nil {
+		return d.BroadcastDKG(ctx, packet.GetDkg())
+	}
 
-	return responseOrError(err)
-}
+	beaconID := packet.Metadata.BeaconID
+	me, err := d.identityForBeacon(beaconID)
+	if err != nil {
+		return nil, err
+	}
 
-func (d *Process) Execute(ctx context.Context, kickoff *drand.StartExecution) (*drand.EmptyResponse, error) {
-	ctx, span := metrics.NewSpan(ctx, "dkg.Execute")
-	defer span.End()
-	beaconID := kickoff.Metadata.BeaconID
+	current, err := d.store.GetCurrent(beaconID)
+	if err != nil {
+		return nil, err
+	}
 
-	err := d.executeAction("DKG execution", beaconID, func(me *drand.Participant, current *DBState) (*DBState, error) {
-		err := d.verifyMessage("StartExecute", kickoff.Metadata, termsFromState(current))
-		if err != nil {
-			return nil, err
-		}
-		return current.Executing(me)
-	})
+	nextState, err := current.Apply(me, packet)
+	if err != nil {
+		return nil, err
+	}
 
+	// we must verify the message against the next state, as the current state upon first proposal will be empty
+	err = d.verifyMessage(packet, packet.Metadata, termsFromState(nextState))
 	if err != nil {
-		d.log.Errorw("There was an error starting the DKG", "beaconID", beaconID, "error", err)
-		return responseOrError(err)
+		return nil, fmt.Errorf("invalid packet signature from %s: %w", packet.Metadata.Address, err)
 	}
 
-	d.log.Infow("DKG execution started successfully", "beaconID", beaconID)
-	dkgConfig, err := d.setupDKG(ctx, beaconID)
+	err = d.store.SaveCurrent(beaconID, nextState)
 	if err != nil {
 		return nil, err
 	}
 
-	d.log.Infow("DKG execution setup successful", "beaconID", beaconID)
+	recipients := util.Concat(nextState.Joining, nextState.Remaining, nextState.Leaving)
+	// we ignore the errors here because it's a best effort gossip
+	// however we can continue with execution
+	_, _ = d.gossip(me, recipients, packet)
+	// we could theoretically ignore when the gossip ends, but due to the mutex we're holding it _could_ lead to a race
+	// condition with future requests
 
-	go func() {
-		time.Sleep(d.config.KickoffGracePeriod)
-		// copy this to avoid any data races with kyber
-		dkgConfigCopy := *dkgConfig
-		err := d.executeAndFinishDKG(ctx, beaconID, dkgConfigCopy)
-		if err != nil {
-			d.log.Errorw("there was an error during the DKG execution!", "beaconID", beaconID, "error", err)
+	if packet.GetExecute() != nil {
+		if err := d.executeDKG(ctx, beaconID); err != nil {
+			return nil, err
 		}
-	}()
+	}
 
-	return responseOrError(err)
+	return &drand.EmptyDKGResponse{}, nil
+}
+
+func commandType(command *drand.DKGCommand) string {
+	switch command.Command.(type) {
```

Comment:
This is a nil pointer dereference on a command==nil, are we sure that can't happen?

Reply:
Have added a check